### ISSUE TITLE
Android 8 and uiautomator2 issues.

### ISSUE THREAD

**Issue by GrzegorzSidwa** [appium/appium-uiautomator2-server#175]
During performing some tests on new devices with android 8 I've noticed that there is a problem with performing them using uiautomator2. At first I thoguht that it is related to appium-dotnet-driver. But I've reproduced the issue on java-client. The behavior is that on some requests server is hanging. I tried both versions, with click, with touch actions. Nothing is working.

**_Java-client code_:**
`DesiredCapabilities capabilities = new DesiredCapabilities();
 capabilities.setCapability(MobileCapabilityType.DEVICE_NAME, "auto");
 capabilities.setCapability(MobileCapabilityType.APP, app.getAbsolutePath());
 capabilities.setCapability(MobileCapabilityType.AUTOMATION_NAME, 
 AutomationName.ANDROID_UIAUTOMATOR2);
 AppiumDriver<MobileElement> driver = new AndroidDriver<>(
                    new URL("http://127.0.0.1:4723/wd/hub"),
                    capabilities);
  MobileElement element = driver.findElementById("android:id/tabs");
  final List<MobileElement> elementsByXPath = element.findElementsByXPath("//*");
  element.findElementsByXPath("//*").stream().filter((x) -> x.getText().equals("CONTROLS")).findFirst().get().click();
`

_**logs from java-client uiautomator2 execution**_
> [HTTP] --> POST /wd/hub/session/5ed3282d-6d05-4ce1-b61c-65e51b56e007/element/160bbe9a-b513-411a-98fe-072e96ee6576/click
[HTTP] {"id":"160bbe9a-b513-411a-98fe-072e96ee6576"}
[debug] [W3C] Calling AppiumDriver.click() with args: ["160bbe9a-b513-411a-98fe-072e96ee6576","5ed3282d-6d05-4ce1-b61c-65e51b56e007"]
[debug] [JSONWP Proxy] Proxying [POST /element/160bbe9a-b513-411a-98fe-072e96ee6576/click] to [POST http://localhost:8200/wd/hub/session/4ad824e3-8852-41ca-922b-c262b248a78d/element/160bbe9a-b513-411a-98fe-072e96ee6576/click] with body: {"element":"160bbe9a-b513-411a-98fe-072e96ee6576"}
[HTTP] <-- POST /wd/hub/session/5ed3282d-6d05-4ce1-b61c-65e51b56e007/element/160bbe9a-b513-411a-98fe-072e96ee6576/click - - ms - -

**_appium-dotnet-driver code_**
`AndroidDriver<AppiumWebElement> driver = new AndroidDriver<AppiumWebElement>(new Uri("http://127.0.0.1:4723/wd/hub"), caps, TimeSpan.FromMinutes(15));
var element = driver.FindElementById(@"android:id/tabs");
var appiumWebElement = element.FindElementsByXPath("//*").FirstOrDefault(x => x.Text == "CONTROLS");
new TouchAction(driver).Press(appiumWebElement).Wait(500).Release().Perform();`

_**logs from appium-dotnet-driver execution**_
> [HTTP] --> POST /wd/hub/session/23a503cc-e4b2-4e48-b750-63e4d362b354/touch/perform
[HTTP] {"actions":[{"action":"press","options":{"element":"dcbad3b6-d315-42f7-a574-8ebb59fead73"}},{"action":"wait","options":{"ms":500}},{"action":"release"}]}
[debug] [W3C] Calling AppiumDriver.performTouch() with args: [[{"action":"press","options":{"element":"dcbad3b6-d315-42f7-a574-8ebb59fead73"}},{"action":"wait","options":{"ms":500}},{"action":"release"}],"23a503cc-e4b2-4e48-b750-63e4d362b354"]
[UiAutomator2] calling get location: dcbad3b6-d315-42f7-a574-8ebb59fead73
[debug] [JSONWP Proxy] Proxying [GET /element/dcbad3b6-d315-42f7-a574-8ebb59fead73/location] to [GET http://localhost:8200/wd/hub/session/b493e649-4d2f-48ca-af88-a3ebacd59f51/element/dcbad3b6-d315-42f7-a574-8ebb59fead73/location] with body: {}

**_Conclusion_**
Exactly the same tests executed on device with Android 7 are working correctly.

**Comment by GrzegorzSidwa** [comment]
Very similar issue has been described here: https://github.com/appium/appium-uiautomator2-server/issues/101

**Comment by mykola-mokhnach** [comment]
@ArturKurowski1 As you can see the code is stuck inside standard Android lib. I don't think that anything could be changed there from Appium side.

**Comment by ArturKurowski1** [comment]
@mykola-mokhnach Yes I noticed it goes into infinite loop  inside UIObject2.getVisibleBounds  since we do not have scrollable element in our test app on that screen , and for some reasonse getting parent is infinite. From Appium and ui-automator2-server everything is fine

**Comment by shuoyoung** [comment]
https://issuetracker.google.com/issues/111312903

**Comment by tzah4748** [comment]
**The Issue:**
So the issue here as mentioned above is that getVisibleBounds is stuck in an endless loop because the parent is never null and is never a scrollable element.
The REAL issue is probably originates from the application itself, i debugged an application with this issue and it seems that eventually getParent will return an element we already iterated on... meaning there is a recursion of parents that none of them is a scrollable element, which causes the loop to be... well, an infinity loop.

**The Solution!...**
NOTE: getting a parent with getParent creates a new Object everytime from the parentId!
Luckily, this is something that can be solved, we just need create a Set that holds all the parentId we already iterated on in the for loop and once we find a parentId that is already inside our Set we can safely break the loop knowing we are probably stuck on the right element.

The concrete solution without changing internal libraries is to actually override them.
I've attached my simple override in a zip file.
[ClickFix.zip](https://github.com/appium/appium-uiautomator2-server/files/5136755/ClickFix.zip)
